ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/CUDAFootstepOptimizer.java

```diff
@@ -0,0 +1,321 @@
+   {
+      cpuGlobalBestIndex = new IntPointer(1);
+      cpuGlobalBestCost = new FloatPointer(1);
+      try (PointerScope ignored = new PointerScope())
+      {
+         blockSize.x(512);
+
+         int resultSize = (searchSpaceDim + blockSize.x() -1) / blockSize.x();
+         FloatPointer cpuBestCosts = new FloatPointer(resultSize);
+         FloatPointer gpuBestCosts = new FloatPointer();
+         CUDATools.mallocAsync(gpuBestCosts, resultSize, cudaStream);
+
+         IntPointer cpuBestIndices = new IntPointer(resultSize);
+         IntPointer gpuBestIndices = new IntPointer();
+         CUDATools.mallocAsync(gpuBestIndices, resultSize, cudaStream);
+
+         LogTools.info("Running block kernel result");
+         // Run the kernel
+         blockSize.x(512);
+         gridSize.x(resultSize);
+         blockResultKernel.withPointer(gpuCosts)
+                          .withInt(searchSpaceDim)
+                          .withPointer(gpuBestCosts)
+                          .withPointer(gpuBestIndices)
+                          .run(cudaStream, gridSize, blockSize, 0);
+
+         CUDATools.memcpyAsync(cpuBestCosts, gpuBestCosts, resultSize, cudaStream);
+         CUDATools.memcpyAsync(cpuBestIndices, gpuBestIndices, resultSize, cudaStream);
+
+         for (int i = 0; i < resultSize; i++)
+            LogTools.info("Best Block Costs: {}, Indices: {}", cpuBestCosts.get(i), cpuBestIndices.get(i));
+
+         searchSpaceDim = resultSize;
+         FloatPointer gpuGlobalBestCost = new FloatPointer();
+         IntPointer gpuGlobalBestIndex = new IntPointer();
+         CUDATools.mallocAsync(gpuGlobalBestIndex, 1, cudaStream);
+         CUDATools.mallocAsync(gpuGlobalBestCost, 1, cudaStream);
+
+         LogTools.info("Running global kernel result");
+         // Run the kernel
+         blockSize.x(resultSize);
+         gridSize.x(1);
+         globalResultKernel.withPointer(gpuBestCosts)
+                           .withPointer(gpuBestIndices)
+                           .withInt(searchSpaceDim)
+                           .withPointer(gpuGlobalBestCost)
+                           .withPointer(gpuGlobalBestIndex)
+                           .run(cudaStream, new dim3(), blockSize, 0);
+
+         CUDATools.memcpyAsync(cpuGlobalBestCost, gpuGlobalBestCost, 1, cudaStream);
+         CUDATools.memcpyAsync(cpuGlobalBestIndex, gpuGlobalBestIndex, 1, cudaStream);
+         cudart.cudaStreamSynchronize(cudaStream);
+
+         LogTools.info("Best Cost: {}, Idx: {}", cpuGlobalBestCost.get(0), cpuGlobalBestIndex.get(0));
+
+         // Release stuff
+         blockSize.close();
+         gridSize.close();
+         cudaFreeAsync(gpuCosts, cudaStream);
+         cudaFreeAsync(cpuGlobalBestCost, cudaStream);
+         cudaFreeAsync(gpuGlobalBestIndex, cudaStream);
+         cudaFreeAsync(gpuBestCosts, cudaStream);
+         cudaFreeAsync(gpuBestIndices, cudaStream);
+      }
+   }
+
+   public void setGpuCosts(float[] costs)
```

Comment:
If this class gets moved to the same package as the test, could we make this method package-private, since it's only used in a test and doesn't need to be public API?
```suggestion
   void setGpuCosts(float[] costs)
```